File: clients/python/baml_lib/_impl/deserializer/deserializer.py
import typing

from .base_deserializer import BaseDeserializer, ITypeDefinition, Diagnostics
from .enum_deserializer import EnumDeserializer
from .object_deserializer import ObjectDeserializer
from .complex_deserializer import (
    ListDeserializer,
    UnionDeserializer,
    OptionalDeserializer,
)
from .type_definition import type_to_definition

from .exports import (
    DefaultDeserializerLUT,
    GeneratedDeserializerLUT,
)

from .raw_wrapper import from_string


T = typing.TypeVar("T")


class Deserializer(typing.Generic[T]):
    __lut: typing.Dict[str, BaseDeserializer[typing.Any]]
    __target_output_type: ITypeDefinition

    def __init__(self, output_target: typing.Type[T]) -> None:
        self.__lut = {}
        self.__target_output_type = type_to_definition(output_target)

    def overload(
        self, name: str, aliases: typing.Dict[str, typing.Optional[str]]
    ) -> None:
        assert name in GeneratedDeserializerLUT, f"Overloading {name} is not allowed."
        default_serializer = GeneratedDeserializerLUT.get(name)
        # default_serializer is a bound method, so we need to get the owning class
        assert default_serializer is not None, f"Could not find owning class for {name}"
        assert name not in self.__lut, f"Overloading {name} twice is not allowed."
        if isinstance(default_serializer, (EnumDeserializer, ObjectDeserializer)):
            deserializer = default_serializer.copy_with_aliases(aliases)
            self.__lut[name] = deserializer
        else:
            assert (
                False
            ), f"Cannot overload {name} with aliases for {type(default_serializer)}"

    def __from_lut(self, dfn: ITypeDefinition) -> BaseDeserializer[T]:
        if dfn["type"] == "List":
            assert "item" in dfn, "List type must have an item"
            return typing.cast(BaseDeserializer[T], ListDeserializer(item=dfn["item"]))
        if dfn["type"] == "Union":
            assert "choices" in dfn, "Union type must have choices"
            return UnionDeserializer(
                *dfn["choices"],
            )
        if dfn["type"] == "Optional":
            assert "item" in dfn, "Optional type must have an item"
            return typing.cast(
                BaseDeserializer[T], OptionalDeserializer(item=dfn["item"])
            )
        if dfn["type"] == "Ref":
            assert "ref" in dfn, "Ref type must have a ref"
            key = dfn["ref"]
            if key.__name__ in self.__lut:
                return self.__lut[key.__name__]
            found = GeneratedDeserializerLUT.get(key.__name__)
            assert found is not None, f"Could not find deserializer for {key.__name__}"
            return found
        found = DefaultDeserializerLUT.get(dfn["type"])
        assert found is not None, f"Could not find deserializer for {dfn['type']}"
        return found

    def from_string(self, s: str) -> T:
        diagnostics = Diagnostics(s)
        raw = from_string(s, diagnostics)
        deserializer = self.__from_lut(self.__target_output_type)
        result = deserializer.coerce(raw, diagnostics, self.__from_lut)
        diagnostics.to_exception()
        return result.as_value

File: clients/ts/src/deserializer/deserializer.ts
import { JSONSchema7, JSONSchema7Definition } from "json-schema";
import { BaseDeserializer } from "./base_deserializer";
import { NoneDeserializer, PrimitiveDeserializer } from "./primitive_deserializer";
import { EnumDeserializer } from "./enum_deserializer";
import { ObjectDeserializer } from "./object_deserializer";
import { fromValue } from "./raw_wrapper/loader";
import { Diagnostics } from "./diagnostics";
import { ListDeserializer, UnionDeserializer } from "./complex_deserializer";

const DefaultDeserializerLUT = {
    None: new NoneDeserializer(),
    bool: new PrimitiveDeserializer(x => x.as_bool(), "Expected bool", 7),
    int: new PrimitiveDeserializer(x => x.as_int(), "Expected int", 8),
    float: new PrimitiveDeserializer(x => x.as_float(), "Expected float", 9),
    string: new PrimitiveDeserializer<string>(x => x.as_str(false), "Expected str", 1),
}

const GeneratedDeserializerLUT = new Map<string, EnumDeserializer<any> | ObjectDeserializer<any>>();


const registerEnumDeserializer = <T extends Record<string, string>>(schema: JSONSchema7, aliases: Record<string, string>) => {
    const deserializer = EnumDeserializer.from_schema<T>(schema, aliases);
    if (GeneratedDeserializerLUT.has(deserializer.name)) {
        console.warn(`Deserializer for ${deserializer.name} already exists. (If you are in development mode using hot-reloading this may be expected.)`);
    }
    GeneratedDeserializerLUT.set(deserializer.name, deserializer);
}

const registerObjectDeserializer = <T extends Record<string, any>>(schema: JSONSchema7, aliases: Record<string, string>) => {
    const deserializer = ObjectDeserializer.from_schema<T>(schema, aliases);
    if (GeneratedDeserializerLUT.has(deserializer.name)) {
        console.warn(`Deserializer for ${deserializer.name} already exists. (If you are in development mode using hot-reloading this may be expected.)`);
    }
    GeneratedDeserializerLUT.set(deserializer.name, deserializer);
}

class Deserializer<T> {
    private overrides: Map<string, EnumDeserializer<any> | ObjectDeserializer<any>> = new Map();

    constructor(private schema: JSONSchema7, private target: JSONSchema7) {
    }

    // on aliases that are scoped to a specific impl, this needs to be specialized for the constructed deserializer instance
    // for that specific impl
    overload(name: string, aliases: Record<string, string>) {
        if (!GeneratedDeserializerLUT.has(name)) {
            throw new Error(`Deserializer for ${name} not found`);
        }

        const overridden = GeneratedDeserializerLUT.get(name)!.copy_with_aliases(aliases);
        if (this.overrides.has(name)) {
            console.warn(`Overload for ${name} already exists. (If you are in development mode using hot-reloading this may be expected.)`);
        }
        this.overrides.set(name, overridden);
    }

    private getInterface(t: JSONSchema7): JSONSchema7 {
        if (t.$ref) {
            const name = t.$ref.split("/").pop();
            if (name === undefined) {
                throw new Error(`Invalid reference`);
            }
            if (this.schema.definitions === undefined) {
                throw new Error(`No definitions found`);
            }

            const schema = this.schema.definitions[name];
            if (schema === undefined) {
                throw new Error(`Definition ${name} not found`);
            }
            if (typeof schema === "boolean") {
                throw new Error(`Definition ${name} is a boolean`);
            }
            return this.getInterface(schema);
        } else {
            return t;
        }
    }

    get_deserializer(_t: JSONSchema7Definition): BaseDeserializer<any> {
        if (typeof _t === "boolean") {
            throw new Error(`Boolean types are not supported`);
        }
        const t = this.getInterface(_t);
        if (t.type === "object" || t.enum) {
            const name = t.title;
            if (name === undefined) {
                throw new Error(`${t.enum ? "Enum" : "Object"} schema must have a title`);
            }
            const deserializer = this.overrides.get(name) ?? GeneratedDeserializerLUT.get(name);
            if (deserializer === undefined) {
                throw new Error(`Deserializer for ${name} not found`);
            }
            return deserializer;
        } else if (t.type === "string") {
            return DefaultDeserializerLUT.string;
        } else if (t.type === "boolean") {
            return DefaultDeserializerLUT.bool;
        } else if (t.type === "integer") {
            return DefaultDeserializerLUT.int;
        } else if (t.type === "number") {
            return DefaultDeserializerLUT.float;
        } else if (t.type === "null") {
            return DefaultDeserializerLUT.None;
        } else if (t.type === "array") {
            if (t.items === undefined) {
                throw new Error(`Array schema must have items`);
            }
            if (Array.isArray(t.items)) {
                throw new Error(`Tuple types are not supported`);
            }
            const item = this.get_deserializer(t.items);
            return new ListDeserializer(item);
        } else if (t.anyOf) {
            const items = t.anyOf.map(x => this.get_deserializer(x));
            return new UnionDeserializer(items);
        } else if (Array.isArray(t.type)) {
            // This is done for optional types
            const items = t.type.map(x => this.get_deserializer({ type: x }));
            return new UnionDeserializer(items);
        } else {
            throw new Error(`Unsupported schema type: ${JSON.stringify(_t)}\n ${t.type}`);
        }

    }

    coerce(value: string): T {
        const d = new Diagnostics(value);
        const raw = fromValue(value, d);

        const deserializer = this.get_deserializer(this.target);
        const response = deserializer.coerce(raw, d, this.get_deserializer.bind(this));
        d.toException();
        return response.as_value;
    }
}

export { Deserializer, registerEnumDeserializer, registerObjectDeserializer }
